### PR TITLE
feat: add extededAddresses field to project struct.

### PR DESCRIPTION
**Pull request by Jeremy-boo**
# Changes
There are different links for different uses of tools, add dependencyAddress field in the project, and support dependent repository address.

update in spec is here:https://github.com/katanomi/spec/pull/226/files 

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] [`spec` PR link](https://github.com/katanomi/spec) included
- [x] Follows the [commit message standard](https://github.com/katanomi/spec/blob/main/CONTRIBUTING.md#commits)
- [ ] Meets the [contributing guidelines](https://github.com/katanomi/spec/blob/main/CONTRIBUTING.md) (including
  functionality, content, code)
- [ ] Test cases with documentation and functionality works as expected using current and related github repos (MUST deploy and check)
- [ ] Release notes block below has been filled in or deleted (only if no user facing changes)

# Release Notes

### CODE REVIEW

File: apis/meta/v1alpha1/labels_annotations.go [hunk 1]
```diff
@@ -135,6 +135,11 @@ const (
 	// GitPRRevisionPrefixes allows git related integrations
 	// define custom PR revision prefixes
 	GitPRRevisionPrefixes = "gitPRRevisionPrefixes"
+
```
Comment (Jeremy-boo) [inline apis/meta/v1alpha1/labels_annotations.go:138]: nexus integreation-class status atrributes 增加的key

File: plugin/client/interface.go [hunk 2]
```diff
@@ -80,6 +80,10 @@ type ResourcePathFormatter interface {
 	GetResourcePathFmt() map[metav1alpha1.ResourcePathScene]string
 	// GetSubResourcePathFmt resource path format
 	GetSubResourcePathFmt() map[metav1alpha1.ResourcePathScene]string
+
```
Comment (Jeremy-boo) [inline plugin/client/interface.go:83]: pkg plugin 插件中新增函数GetPYPIResourcePathFmt(), 该函数主要是插件里面可以自定义 自己subResource 的fmt

File: apis/meta/v1alpha1/project_types.go [hunk 3]
```diff
@@ -172,6 +183,10 @@ type ProjectSpec struct {
 	// +optional
 	Access *duckv1.Addressable `json:"access,omitempty"`
 
+	// ExtendAddress stores the tool extend url which has different repository than address.
+	// +optional
+	ExtendAddress map[ExtendAddressKey]*duckv1.Addressable `json:"extendAddress,omitempty"`
```
Comment (danielfbm) [inline apis/meta/v1alpha1/project_types.go:188]: `extendAddress` -> `extendedAddresses `

File: apis/meta/v1alpha1/project_types.go [hunk 4]
```diff
@@ -147,6 +147,17 @@ const (
...
 
+// ExtendAddressKey The name of the plugin extension url, only dependencyAddress is supported for now.
+// TODO: add more key for the plugin extension url.
+type ExtendAddressKey string
+
+const (
+	// DependencyAddressExtendKey describes the dependency repository address provided by the plugin
+	// If the tool has an independent dependency repository address,
+	// use this field to indicate, if not, take the same value as address
+	DependencyAddressExtendKey ExtendAddressKey = "dependencyAddress"
```
Comment (danielfbm) [inline apis/meta/v1alpha1/project_types.go:158]: `dependency` 通过上下文是不是就可以知道是address呢？这个 ExtendedAddresses 会有别的非address的东西吗？

File: apis/meta/v1alpha1/project_types.go [hunk 5]
```diff
@@ -147,6 +147,17 @@ const (
 	// TODO: add more subtypes
 )
 
+// ExtendAddressKey The name of the plugin extension url, only dependencyAddress is supported for now.
+// TODO: add more key for the plugin extension url.
+type ExtendAddressKey string
```
Comment (danielfbm) [inline apis/meta/v1alpha1/project_types.go:152]: `ExtendedAddress`

File: plugin/client/interface.go [hunk 6]
```diff
@@ -80,6 +80,10 @@ type ResourcePathFormatter interface {
...
+	// GetPYPIResourcePathFmt return a list of pypi resource fmt that can be used in get projects list api.
+	// TODO: we need remove this func after find better way to add suffix to pypi dependency url.
+	GetPYPIResourcePathFmt() map[metav1alpha1.ProjectSubType]string
```
Comment (danielfbm) [inline plugin/client/interface.go:86]: 临时的话是不是直接从IntegrationClass取数据就行了？为啥要改client？

**Review by danielfbm — CHANGES_REQUESTED** [review]
见评论

File: plugin/client/interface.go (hunk 6 above)
Comment (Jeremy-boo) [reply plugin/client/interface.go:86]: 这个client 里面的方法主要是提供给插件 做初始化用的，插件实现GetPYPIResourcePathFmt 方法，设置初始化的数据，然后integration_controller 通过这个办法 获取到插件里面设置的初始化数据，然后给它设置到IntegrationClass 里面的atrributes里面去。

File: apis/meta/v1alpha1/project_types.go (hunk 4 above)
Comment (Jeremy-boo) [reply apis/meta/v1alpha1/project_types.go:158]: ExtendedAddresses 这里讨论的是不会放非address 的东西进来的，后续字段扩展开来可能会有

releaseAddress，httpCloneAddress 这些

File: apis/meta/v1alpha1/labels_annotations.go [hunk 7]
```diff
@@ -135,6 +135,11 @@ const (
...
+	//  PYPISubResourceSuffixKey Add the corresponding suffix to the specified subResource which for pypi.
+	// TODO: In the future, we need to adopt a better way to add
+	// an extended dependency repository address to the plug-in
+	PYPISubResourceSuffixKey = "pypiSubResourceSuffix"
```
[the remark this answers is not in the document]
Comment (Jeremy-boo) [reply apis/meta/v1alpha1/labels_annotations.go:142]: 嗯啦，这样看代码能够更联想起来。

**Review by danielfbm — APPROVED** [review]

**Pull request merged**